src/Interpreters/findGoodReordering.cpp

```diff
@@ -0,0 +1,246 @@
+{
+    /// We do "removal" of indexes from the data structures used by greedy() by
+    ///     1) copying the element at the back of the data structure, at index n-1, to the index to be removed.
+    ///     2) decrementing n.
+    for (size_t i = 0; i < elems_to_remove.size(); ++i)
+    {
+        size_t idx_to_remove = elems_to_remove[i];
+
+        if (i > 0)
+            idx_to_remove = std::find(indexes.begin(), indexes.end(), idx_to_remove) - indexes.begin();
+
+        //node idx_to_remove is now a part of the join tree, update selectivities for the remaining nodes.
+        for (size_t k = 0; k < n; k++)
+        {
+            selectivity_to_joined[k] *= selectivities(idx_to_remove, k);
+        }
+
+        //move last node (index: n-1) to the place of selected node (index: idx_to_remove).
+        if (idx_to_remove != n-1)
+        {
+            for (size_t k = 0; k < n; k++)
+            {
+                selectivities(k, idx_to_remove) = selectivities(n-1, k);
+                selectivities(idx_to_remove, k) = selectivities(n-1, k);
+            }
+            selectivity_to_joined[idx_to_remove] = selectivity_to_joined[n-1];
+            self_selectivities[idx_to_remove] = self_selectivities[n-1];
+            sizes[idx_to_remove] = sizes[n-1];
+            indexes[idx_to_remove] = indexes[n-1];
+        }
+        n -= 1;
+    }
+    return n;
+}
+
+static Float64 calculateCost(Float64 size_left, Float64 size_right, Float64 num_cols_right, Float64 sel_left, Float64 sel_right)
+{
+    /// The constants here come from a linear regression fitted to predict runtimes of joins of two tables.
+    /// The regression fits the vector coefs to sum(X*coefs)=y
+    /// y = join runtime in microseconds
+    /// X = [    size_left,  size_right, size_left*num_cols_left, size_right*num_cols_right, size_left*sel_left, size_right*sel_right]
+    /// coefs = [0.01709865, 0.        , 0.                     , 0.01674002,                0.06704978,          0.3567766 ]
+    /// The regression was constrained to have non-negative coefficients. All examples in the training data used hash joins.
+    /// It got an R2 = 0.965 on the training data (which had 320 samples).
+    /// Inputs to the regression are clearly correlated, and coefficients may not map to any real characteristics of how joins work.
+    return size_left * 0.01709865 +
+           num_cols_right * size_right * 0.01674002 +
+           size_left * sel_left * 0.06704978 +
+           size_right * sel_right * 0.3567766;
+}
+
+static void updateCost(Float64 & total_cost, Float64 & size_left, Float64 size_right, Float64 num_cols_right, Float64 sel_left, Float64 sel_right)
+{
+    /// take into account that lhs costs less than rhs
+    total_cost += calculateCost(size_left, size_right, num_cols_right, sel_left, sel_right);
+    size_left *= size_right * sel_left * sel_right;
+}
+
+static Float64 greedy(const size_t N, /// Number of remaining tables to join
```

Comment:
Won't be the same result if we just sort tables by size? Could you please add a benchmark for it?

Reply:
The cost is highly dependent on the graph of join keys (i.e. which determines which tables can be inner joined, vs which tables must be cross joined).

Simply sorting by size won't work in general. For example, consider the query:

SELECT * FROM t1, t2, t3 WHERE t1.id = t3.id AND t2.oid = t3.oid;

Let's say we have table sizes:

t1: 1 billion rows
t2: 1 million rows
t3: 1 thousand rows

Here we don't have a join key between t1 and t2. Simply ordering the tables by size would then produce a cross join between t1 and t2, which would have a quadrillion rows (literally, 1M*1B=10^15). The quadrillion row table would then have to be inner joined with t3. On the other hand, the ordering t1, t3, t2 would allow inner joins all the way, producing much smaller intermediate tables.

Reply:
Am i right, that you still do join
`(T1 JOIN T3) JOIN T2`

Or it something like:

`T1 JOIN (T3 JOIN T2)`

Reply:
You're right, it only finds left deep trees. This is mostly because otherwise, it would require a rewrite of the JoinToSubquery pass.